File: Python/Base/TestGraph2.py
import numpy as np
import time

def TestGraph2(Test, A, B, X1, cluster, center_vector):
    eta = 0.000001
    Test_num, test_attribute_num = Test.shape
    datatest = Test[:, :-1]
    nhantest = Test[:, -1]

    start_time_train = time.time()

    # Initialize an empty array for cluster assignments
    X2 = np.zeros_like(datatest, dtype=int)

    # Calculate cluster assignment for each test point
    for j in range(test_attribute_num - 1):
        w = center_vector[j]
        for i in range(Test_num):
            d = np.ones(len(w)) * 1000  # Set all distances initially to a large value
            for l in range(len(w)):
                d[l] = datatest[i, j] - w[l]
            dmin = np.min(d)
            X2[i, j] = np.argmin(d)  # Find index of the minimum distance

    T = X2  # Use cluster assignment as T

    ruleList1 = X1
    te = T
    a1, a2 = te.shape
    r1, r2 = ruleList1.shape

    # Initialize matrices C1, C2, and C3 for storing results
    C1 = np.zeros((a1, a2 - 2))
    C2 = np.zeros((a1, a2 - 2))
    C3 = np.zeros((a1, a2 - 2))

    # Compute C1, C2, C3
    for i in range(a1):
        for l in range(a2 - 2):
            for h in range(l + 1, a2 - 1):
                dem1 = dem2 = dem3 = 0
                for t in range(r1):
                    if te[i, l] == ruleList1[t, l] and te[i, h] == ruleList1[t, h]:
                        if ruleList1[t, r2 - 1] == 1:
                            dem1 += B[t, 0]
                        elif ruleList1[t, r2 - 1] == 2:
                            dem2 += B[t, 0]
                        else:
                            dem3 += B[t, 0]
                C1[i, l] = dem1
                C2[i, l] = dem2
                C3[i, l] = dem3

    minC1 = np.min(C1, axis=1)
    maxC1 = np.max(C1, axis=1)
    D1 = minC1 + maxC1

    minC2 = np.min(C2, axis=1)
    maxC2 = np.max(C2, axis=1)
    D2 = minC2 + maxC2

    minC3 = np.min(C3, axis=1)
    maxC3 = np.max(C3, axis=1)
    D3 = minC3 + maxC3

    nhan = np.zeros(a1)

    # Classification based on distances D1, D2, D3
    for k in range(a1):
        if D1[k] > D2[k] and D1[k] > D3[k]:
            nhan[k] = 0
        elif D3[k] > D2[k] and D3[k] > D1[k]:
            nhan[k] = 1
        else:
            nhan[k] = 2

    end_time_train = time.time()
    Time = end_time_train - start_time_train

    # Compute accuracy
    temp = nhan
    train_output = nhantest
    dem = np.sum(temp == train_output)
    
    Accuracy = dem / len(temp)

    return Accuracy, Time

File: Matlab/TestGraph2.m

function [Accuracy] = TestGraph2(Test,A,B,X1,cluster,center_vector)
% var_data = [train_data(:,1) var_data];
eta=0.000001;
[Test_num test_attribute_num] = size(Test);
datatest=Test(:,1:test_attribute_num-1);
nhantest=Test(:,test_attribute_num);

start_time_train = cputime;
for j=1:test_attribute_num-1
    w=center_vector{j};
    for i=1:Test_num
        for l=1:cluster(j)
            d(l)=1000;
            d(l)=datatest(i,j)-w(l);
        end
        dmin=min(d);
        for l=1:cluster(j)
            if d(l)==dmin
                X2(i,j)=l;
            end
        end
    end
end
T=X2;
% T=X1
% T=temp;

%  [a,b]=size(V);
% T=[2 1 1 1 1 1;1 2 1 1 3 1;3 2 2 2 3 1];
ruleList1=X1;
te=T;
%  nhan1=[1 2 1 1 2 2];
% te=ruleList1(:,1:9);
size(A);
size(B);
% nhan1=ruleList1(:,10)';
[a1,a2]=size(te);
[r1,r2]=size(ruleList1);
ruleList1;

%Tính C
for i=1: a1
    j=1;
    for l=1:r2-2
        for h=l+1:r2-1
            dem1=0;
            dem2=0;
            dem3=0;
            for t=1:r1
                if (te(i,l)==ruleList1(t,l))&(te(i,h)==ruleList1(t,h))
                    if (ruleList1(t,r2)==1)
                        dem1=dem1+B(t,j);
                    else
                        if (ruleList1(t,r2)==2)
                            dem2=dem2+B(t,j);
                        else
                            dem3=dem3+B(t,j);
                        end
                    end
                end
            end
            C1(i,j)=dem1;
            C2(i,j)=dem2;
            C3(i,j)=dem3;
            j=j+1;
        end
    end
end
C1;
C2;
C3;
% for j=1:a2
%      for k=1:a1
%             tong=0;
%             for t=1:b1
%                 if (te(k,j)==ruleList1(t,j))&(ruleList1(t,b2)==0)
%                     tong=tong+B(t,j);
%                 end
%             end
%             C1(k,j)=tong;
%     end
% end
% C1;
% 
 minC1=min(C1');
 maxC1=max(C1');
 
D1=minC1+maxC1;

% D1=minC1;
% for j=1:a2
%      for k=1:a1
%             tong=0;
%             for t=1:b1
%                 if (te(k,j)==ruleList1(t,j))&(ruleList1(t,b2)==1)
%                     tong=tong+B(t,j);
%                 end
%             end
%             C2(k,j)=tong;
%     end
% end
D1;
minC2=min(C2');
maxC2=max(C2');
D2=minC2+maxC2;
D2;
minC3=min(C3');
maxC3=max(C3');
D3=minC3+maxC3;
% D2=minC2
for k=1:a1
    if (D1(k)>D2(k))&(D1(k)>D3(k))
        nhan(k)=0;
    else
        if (D3(k)>D2(k))&(D3(k)>D1(k))
            nhan(k)=1;
        else
            nhan(k)=2;
        end
    end
end
nhan;
end_time_train = cputime;
Time =end_time_train - start_time_train;
nhan;
temp=nhan;
train_output=nhantest;
dem=0;
size(temp);
size(train_output);
for i=1:size(temp')
    if temp(i)==train_output(i)
        dem=dem+1;
    end
end
[a111,b111]=size(temp');
Accuracy=dem/a111;
% TN=0;
% TP=0;
% FN=0;
% FP=0;
% for i=1:size(temp')
%     if (temp(i)==train_output(i))& (temp(i)==0)
%         TP=TP+1;
%     end
%     if (temp(i)==train_output(i))& (temp(i)==1)
%         TN=TN+1;
%     end
%     if (train_output(i)==0)& (temp(i)==0)
%         FN=FN+1;
%     end
%     if (train_output(i)==1)& (temp(i)==0)
%         FP=FP+1;
%     end
% end
% 
% Accuracy=(TN+TP)/(TN+TP+FN+FP);
% 
% Recall=TP/(TP+FN);
% 
% Precision=TP/(TP+FP);




